File: sim/core/bulk/stage.go
package bulk

import (
	"log"
	"math"
	"runtime"
	"time"

	"github.com/wowsims/tbc/sim/core"
	"github.com/wowsims/tbc/sim/core/proto"
	"github.com/wowsims/tbc/sim/core/simsignals"
)

type BulkSimStageConfig struct {
	Stage              proto.BulkSimStage
	MinIterations      int32
	TargetErrorPct     float64
	MinSurvivors       int
	MaxSurvivors       int
	CullingCoefficient float64
	UseConcurrentSim   bool
}

var bulkSimStageConfigs = []BulkSimStageConfig{
	{
		Stage:              proto.BulkSimStage_BulkSimStageLow,
		MinIterations:      100,
		TargetErrorPct:     1,
		MinSurvivors:       20,
		MaxSurvivors:       100,
		CullingCoefficient: bulkSimCullingCoefficient,
	},
	{
		Stage:              proto.BulkSimStage_BulkSimStageMedium,
		MinIterations:      1000,
		TargetErrorPct:     0.2,
		MinSurvivors:       5,
		MaxSurvivors:       25,
		CullingCoefficient: bulkSimCullingCoefficient,
	},
	{
		Stage:            proto.BulkSimStage_BulkSimStageHigh,
		MinIterations:    1000,
		TargetErrorPct:   0.05,
		UseConcurrentSim: true,
	},
}

type BulkSimStageResult struct {
	Baseline   *BulkSimCandidateResult
	Results    []*BulkSimCandidateResult
	Iterations int32
	Metrics    *proto.BulkSimStageMetrics
}

func shouldRunBulkSimStage(config BulkSimStageConfig, candidateCount int) bool {
	maxSurvivors := getBulkSimStageMaxSurvivors(config, candidateCount)
	return maxSurvivors == 0 || candidateCount > maxSurvivors || candidateCount < bulkSimMinCombinations && config.Stage == proto.BulkSimStage_BulkSimStageHigh
}

func GetBulkSimStageConcurrency(request *proto.BulkSimRequest, config BulkSimStageConfig) int {
	if config.UseConcurrentSim {
		return 1
	}
	if request.BaseRequest.SimOptions.IsTest {
		return 3
	}
	if config.Stage == proto.BulkSimStage_BulkSimStageLow {
		return runtime.NumCPU() * bulkSimLowStageConcurrencyFactor
	}
	return runtime.NumCPU()
}

// Runs one low/medium/high refinement stage. Each stage first probes the
// baseline to estimate variance, then uses that variance to choose a
// per-candidate iteration count before simming every candidate. After the first
// pass, the stage may add more iterations if the observed error is still above
// the configured target.
//
// Iterations already simmed by the previous stage are carried over instead of
// being re-run: the stage only sims the delta with a seed offset and merges. That
// is exactly equivalent to running the full count from scratch, because iteration
// N always uses seed RandomSeed+N.
func runBulkSimStage(request *proto.BulkSimRequest, candidates []BulkSimCandidate, config BulkSimStageConfig, carry *bulkSimStageCarryOver, progress chan *proto.ProgressMetrics, signals simsignals.Signals) BulkSimStageResult {
	startedAt := time.Now()
	minIterations := getBulkSimStageMinIterations(request.HighStageIterations, config)
	concurrency := GetBulkSimStageConcurrency(request, config)
	concurrency = max(1, min(concurrency, len(candidates)))
	if !carry.covers(candidates) {
		carry = nil
	}
	carriedIterations := carry.carriedIterations()
	log.Printf("[Bulk Sim] %s", formatBulkSimStageStart(config, len(candidates), concurrency, minIterations, carriedIterations))
	maxBaselineSims := 2
	maxTotalSims := len(candidates) + maxBaselineSims
	probeDelta := max(0, minIterations-carriedIterations)
	probeEmitter := bulkSimStageProgressEmitter{progress: progress, stage: config.Stage, totalSims: maxTotalSims, totalIterations: int32(maxTotalSims) * probeDelta}
	probeEmitter.report(0, 0, 0)

	// Run the baseline gear up to the stage minimum to estimate DPS variance.
	// That variance is used to calculate how many iterations are needed for the
	// stage target error; user-provided high-stage iterations are treated as a
	// floor and may be raised if the probe shows more iterations are required.
	baselineProbe := carry.baselineResult()
	probeIterations := carriedIterations
	if probeDelta > 0 {
		baselineProbe = runBulkSimBaselineSegment(request, config, baselineProbe, probeDelta, carriedIterations, probeEmitter, 0, 0, signals)
		if baselineProbe.Error != nil {
			return BulkSimStageResult{Baseline: baselineProbe}
		}
		probeIterations = minIterations
	}
	probeEmitter.report(1, probeDelta, baselineProbe.DpsMetrics.Avg)

	iterations := max(getBulkSimStageIterations(request, config, baselineProbe.DpsMetrics, len(candidates)), probeIterations)
	baselineDelta := iterations - probeIterations
	candidateDelta := iterations - carriedIterations
	probeSims := core.TernaryInt(probeDelta > 0, 1, 0)
	baselineSims := probeSims + core.TernaryInt(baselineDelta > 0, 1, 0)
	candidateSims := core.TernaryInt(candidateDelta > 0, len(candidates), 0)
	totalSims := candidateSims + baselineSims
	completedBaselineIterations := probeDelta
	baseline := baselineProbe

	emitter := bulkSimStageProgressEmitter{progress: progress, stage: config.Stage, totalSims: totalSims, totalIterations: int32(candidateSims)*candidateDelta + probeDelta + baselineDelta}
	emitter.report(probeSims, completedBaselineIterations, baselineProbe.DpsMetrics.Avg)
	if baselineDelta > 0 {
		baseline = runBulkSimBaselineSegment(request, config, baselineProbe, baselineDelta, probeIterations, emitter, probeSims, probeDelta, signals)
		if baseline.Error != nil {
			return BulkSimStageResult{Baseline: baseline, Iterations: iterations}
		}
		completedBaselineIterations = probeDelta + baselineDelta
		emitter.report(baselineSims, completedBaselineIterations, baseline.DpsMetrics.Avg)
	}

	// Everything this stage needs is already covered by the carried iterations.
	if candidateDelta <= 0 {
		collected := carry.results(candidates)
		return BulkSimStageResult{
			Baseline:   baseline,
			Results:    collected,
			Iterations: iterations,
			Metrics:    bulkSimStageMetrics(config, candidates, collected, baseline, iterations, concurrency, startedAt),
		}
	}

	collected := runBulkSimCandidateBatch(request, candidates, config, bulkSimCandidateBatchConfig{
		iterations:              candidateDelta,
		seedOffset:              carriedIterations,
		concurrency:             concurrency,
		completedSimsBase:       baselineSims,
		completedIterationsBase: completedBaselineIterations,
		emitter:                 emitter,
		carried:                 carry,
	}, signals)
	baseline, collected, iterations = adaptBulkSimStageIterations(request, candidates, config, progress, signals, concurrency, baseline, collected, iterations)
	if baseline.Error != nil {
		return BulkSimStageResult{Baseline: baseline, Results: collected, Iterations: iterations}
	}

	return BulkSimStageResult{
		Baseline:   baseline,
		Results:    collected,
		Iterations: iterations,
		Metrics:    bulkSimStageMetrics(config, candidates, collected, baseline, iterations, concurrency, startedAt),
	}
}

// Runs `iterations` of the baseline gear at `seedOffset` and folds the result into
// `carried` (nil when there is nothing carried yet). Progress is reported as
// completedSimsBase sims done plus this segment's iterations on top of
// completedIterationsBase. An errored segment is returned as-is for the caller to check.
func runBulkSimBaselineSegment(request *proto.BulkSimRequest, config BulkSimStageConfig, carried *BulkSimCandidateResult, iterations int32, seedOffset int32, emitter bulkSimStageProgressEmitter, completedSimsBase int, completedIterationsBase int32, signals simsignals.Signals) *BulkSimCandidateResult {
	segment := runSingleBulkSimCandidate(request, BulkSimCandidate{Index: -1, Gear: getBulkSimBaselineGear(request)}, iterations, seedOffset, signals, config.UseConcurrentSim, func(progressMetrics *proto.ProgressMetrics) {
		if progressMetrics.TotalIterations == 0 {
			return
		}
		emitter.report(completedSimsBase, completedIterationsBase+min(progressMetrics.CompletedIterations, iterations), progressMetrics.Dps)
	})
	if segment.Error != nil {
		return segment
	}
	return mergeBulkSimCandidateResults(carried, segment)
}

func bulkSimStageMetrics(config BulkSimStageConfig, candidates []BulkSimCandidate, results []*BulkSimCandidateResult, baseline *BulkSimCandidateResult, iterations int32, concurrency int, startedAt time.Time) *proto.BulkSimStageMetrics {
	return &proto.BulkSimStageMetrics{
		Stage:               config.Stage,
		InputGearSets:       int32(len(candidates)),
		Survivors:           int32(len(results)),
		Iterations:          iterations,
		Concurrency:         int32(concurrency),
		DurationSeconds:     time.Since(startedAt).Seconds(),
		TargetErrorPct:      config.TargetErrorPct,
		ObservedErrorPct:    bulkSimObservedStageErrorPct(baseline, results, iterations, len(candidates)),
		BaselineAvgDps:      baseline.DpsMetrics.Avg,
		BestCandidateAvgDps: bestBulkSimDps(results),
	}
}

func getBulkSimStageMinIterations(highStageIterations int32, config BulkSimStageConfig) int32 {
	if config.Stage == proto.BulkSimStage_BulkSimStageHigh && highStageIterations > 0 {
		return highStageIterations
	}
	return config.MinIterations
}

func getBulkSimStageIterations(request *proto.BulkSimRequest, config BulkSimStageConfig, baselineMetrics *proto.DistributionMetrics, candidateCount int) int32 {
	minIterations := getBulkSimStageMinIterations(request.HighStageIterations, config)
	// The user-defined high-stage iteration count is a floor, not a cap. Every
	// stage still uses enough iterations to satisfy its target error when needed.
	targetIterations := getBulkSimTargetIterations(config.TargetErrorPct, baselineMetrics, candidateCount)
	return max(minIterations, targetIterations)
}

func getBulkSimStageMaxSurvivors(config BulkSimStageConfig, candidateCount int) int {
	if config.MaxSurvivors == 0 {
		return config.MaxSurvivors
	}

	var scaleReference int
	switch config.Stage {
	case proto.BulkSimStage_BulkSimStageLow:
		scaleReference = bulkSimLowStageSurvivorScaleReference
	case proto.BulkSimStage_BulkSimStageMedium:
		scaleReference = bulkSimMediumStageSurvivorScaleReference
	default:
		return config.MaxSurvivors
	}

	if candidateCount <= scaleReference {
		return config.MaxSurvivors
	}

	scale := math.Sqrt(float64(candidateCount) / float64(scaleReference))
	return max(config.MaxSurvivors, int(math.Ceil(float64(config.MaxSurvivors)*scale)))
}

// Adds bounded extra iterations when the completed stage missed its target
// error. Extra sims use seed offsets and are merged into the existing metrics,
// avoiding a full rerun while still reducing standard error for the same
// baseline/candidate set.
func adaptBulkSimStageIterations(request *proto.BulkSimRequest, candidates []BulkSimCandidate, config BulkSimStageConfig, progress chan *proto.ProgressMetrics, signals simsignals.Signals, concurrency int, baseline *BulkSimCandidateResult, results []*BulkSimCandidateResult, iterations int32) (*BulkSimCandidateResult, []*BulkSimCandidateResult, int32) {
	maxAdaptiveIterations := int32(math.Ceil(float64(iterations) * bulkSimAdaptiveMaxIterationMultiplier))
	for adaptivePass := 1; adaptivePass <= bulkSimMaxAdaptivePasses; adaptivePass++ {
		if signals.Abort.IsTriggered() || hasBulkSimStageError(baseline, results) {
			return baseline, results, iterations
		}

		observedErrorPct := bulkSimObservedStageErrorPct(baseline, results, iterations, len(candidates))
		if observedErrorPct <= config.TargetErrorPct {
			return baseline, results, iterations
		}

		targetIterations := getBulkSimStageTargetIterations(config.TargetErrorPct, baseline, results, len(candidates))
		targetIterations = min(maxAdaptiveIterations, max(iterations+1, targetIterations))
		if targetIterations <= iterations {
			return baseline, results, iterations
		}

		additionalIterations := targetIterations - iterations
		log.Printf("[Bulk Sim] - Stage: %s - Adaptive pass %d\nResults:\n  Current iterations: %d\n  Additional iterations: %d\n  Target iterations: %d\n  Target error: %.2f%%\n  Observed error: %.2f%%", bulkSimStageLogName(config.Stage), adaptivePass, iterations, additionalIterations, targetIterations, config.TargetErrorPct, observedErrorPct)
		baseline, results = rerunBulkSimStageAdditionalIterations(request, candidates, config, progress, signals, concurrency, baseline, results, iterations, additionalIterations)
		iterations = targetIterations
	}
	return baseline, results, iterations
}

// Runs only the delta required by an adaptive pass. The seed offset prevents
// reusing the same random sequence as the previous pass, and the returned
// metrics are merged with the existing stage results by candidate index.
func rerunBulkSimStageAdditionalIterations(request *proto.BulkSimRequest, candidates []BulkSimCandidate, config BulkSimStageConfig, progress chan *proto.ProgressMetrics, signals simsignals.Signals, concurrency int, baseline *BulkSimCandidateResult, results []*BulkSimCandidateResult, currentIterations int32, additionalIterations int32) (*BulkSimCandidateResult, []*BulkSimCandidateResult) {
	totalSims := len(candidates) + 1
	emitter := bulkSimStageProgressEmitter{progress: progress, stage: config.Stage, totalSims: totalSims, totalIterations: int32(totalSims) * additionalIterations}
	emitter.report(0, 0, 0)

	baselineExtra := runBulkSimBaselineSegment(request, config, baseline, additionalIterations, currentIterations, emitter, 0, 0, signals)
	if baselineExtra.Error != nil {
		return baselineExtra, results
	}
	baseline = baselineExtra
	emitter.report(1, additionalIterations, baseline.DpsMetrics.Avg)

	// The pass so far is the carry-over for this batch, so the extra iterations are
	// merged into each candidate as they complete - exactly like a stage carrying its
	// predecessor's iterations.
	return baseline, runBulkSimCandidateBatch(request, candidates, config, bulkSimCandidateBatchConfig{
		iterations:              additionalIterations,
		seedOffset:              currentIterations,
		concurrency:             concurrency,
		completedSimsBase:       1,
		completedIterationsBase: additionalIterations,
		emitter:                 emitter,
		carried:                 bulkSimCarryOverFromResults(currentIterations, baseline, results),
	}, signals)
}

File: ui/core/wasm/bulk_sim/stage.ts
import { BulkSimRequest, BulkSimStage, BulkSimStageMetrics, DistributionMetrics, ErrorOutcome } from '../../proto/api';
import { SimSignals } from '../../sim_signal_manager';
import { isDevMode } from '../../utils';
import { WorkerPool, WorkerProgressCallback } from '../../worker_pool';
import { BulkSimCandidateTransport, runBulkSimCandidateBatchOnWorkers, runSingleBulkSimCandidate } from './batch';
import { ConcurrentBulkSimStageCarryOver, bulkSimCarriedResults, bulkSimCarryOverCovers } from './carry_over';
import {
	BULK_SIM_ADAPTIVE_MAX_ITERATION_MULTIPLIER,
	BULK_SIM_CULLING_COEFFICIENT,
	BULK_SIM_LOW_STAGE_SURVIVOR_SCALE_REFERENCE,
	BULK_SIM_MAX_ADAPTIVE_PASSES,
	BULK_SIM_MEDIUM_STAGE_SURVIVOR_SCALE_REFERENCE,
	BULK_SIM_MIN_COMBINATIONS,
} from './constants';
import { getBulkSimBaselineGear } from './index';
import { hasBulkSimStageError, mergeBulkSimCandidateResults } from './merge';
import { BulkSimStageProgressEmitter, bulkSimStageLogName, formatBulkSimStageStart, makeBulkSimStageProgressEmitter } from './progress';
import { bulkSimObservedStageErrorPct, getBulkSimStageTargetIterations, getBulkSimTargetIterations, topBulkSimResults } from './statistics';
import { ConcurrentBulkSimCandidate, ConcurrentBulkSimCandidateResult, ConcurrentBulkSimStageConfig, ConcurrentBulkSimStageResult } from './types';

export const bulkSimStageConfigs: ConcurrentBulkSimStageConfig[] = [
	{
		stage: BulkSimStage.BulkSimStageLow,
		minIterations: 100,
		targetErrorPct: 1,
		minSurvivors: 20,
		maxSurvivors: 100,
		cullingCoefficient: BULK_SIM_CULLING_COEFFICIENT,
	},
	{
		stage: BulkSimStage.BulkSimStageMedium,
		minIterations: 1000,
		targetErrorPct: 0.2,
		minSurvivors: 5,
		maxSurvivors: 25,
		cullingCoefficient: BULK_SIM_CULLING_COEFFICIENT,
	},
	{
		stage: BulkSimStage.BulkSimStageHigh,
		minIterations: 1000,
		targetErrorPct: 0.05,
	},
];

// Scales the survivor cap for large candidate sets, mirroring
// getBulkSimStageMaxSurvivors in sim/core/bulk/stage.go. Returns undefined for
// uncapped stages.
export const getBulkSimStageMaxSurvivors = (config: ConcurrentBulkSimStageConfig, candidateCount: number): number | undefined => {
	if (config.maxSurvivors === undefined) return undefined;

	let scaleReference: number;
	switch (config.stage) {
		case BulkSimStage.BulkSimStageLow:
			scaleReference = BULK_SIM_LOW_STAGE_SURVIVOR_SCALE_REFERENCE;
			break;
		case BulkSimStage.BulkSimStageMedium:
			scaleReference = BULK_SIM_MEDIUM_STAGE_SURVIVOR_SCALE_REFERENCE;
			break;
		default:
			return config.maxSurvivors;
	}

	if (candidateCount <= scaleReference) return config.maxSurvivors;

	const scale = Math.sqrt(candidateCount / scaleReference);
	return Math.max(config.maxSurvivors, Math.ceil(config.maxSurvivors * scale));
};

export const shouldRunBulkSimStage = (config: ConcurrentBulkSimStageConfig, candidateCount: number): boolean => {
	const maxSurvivors = getBulkSimStageMaxSurvivors(config, candidateCount);
	return (
		maxSurvivors === undefined ||
		candidateCount > maxSurvivors ||
		(candidateCount < BULK_SIM_MIN_COMBINATIONS && config.stage == BulkSimStage.BulkSimStageHigh)
	);
};

export const getBulkSimStageMinIterations = (request: BulkSimRequest, config: ConcurrentBulkSimStageConfig): number => {
	if (config.stage == BulkSimStage.BulkSimStageHigh && request.highStageIterations > 0) {
		return request.highStageIterations;
	}
	return config.minIterations ?? request.highStageIterations;
};

const getBulkSimStageIterations = (
	request: BulkSimRequest,
	config: ConcurrentBulkSimStageConfig,
	baselineMetrics: DistributionMetrics | undefined,
	candidateCount: number,
): number => {
	const minIterations = getBulkSimStageMinIterations(request, config);
	const targetIterations = getBulkSimTargetIterations(config.targetErrorPct, baselineMetrics, candidateCount);
	return Math.max(minIterations, targetIterations);
};

// Runs `iterations` of the baseline gear at `seedOffset` and folds the result into
// `carried` (undefined when there is nothing carried yet). Progress is reported as
// completedSimsBase sims done plus this segment's iterations on top of
// completedIterationsBase. An errored segment is returned as-is for the caller to check.
const runBulkSimBaselineSegment = async (
	request: BulkSimRequest,
	carried: ConcurrentBulkSimCandidateResult | undefined,
	iterations: number,
	seedOffset: number,
	workerPool: WorkerPool,
	emitter: BulkSimStageProgressEmitter,
	completedSimsBase: number,
	completedIterationsBase: number,
	signals: SimSignals,
): Promise<ConcurrentBulkSimCandidateResult> => {
	const segment = await runSingleBulkSimCandidate(
		request,
		{ index: -1, gear: getBulkSimBaselineGear(request) },
		iterations,
		workerPool,
		signals,
		BulkSimCandidateTransport.SplitAcrossWorkers,
		progressMetrics => {
			if (progressMetrics.totalIterations == 0) return;
			emitter.report(completedSimsBase, completedIterationsBase + Math.min(progressMetrics.completedIterations, iterations), progressMetrics.dps);
		},
		seedOffset,
	);
	if (segment.error) return segment;
	return mergeBulkSimCandidateResults(carried, segment);
};

const adaptConcurrentBulkSimStageIterations = async (
	request: BulkSimRequest,
	candidates: ConcurrentBulkSimCandidate[],
	config: ConcurrentBulkSimStageConfig,
	workerPool: WorkerPool,
	onProgress: WorkerProgressCallback,
	signals: SimSignals,
	baseline: ConcurrentBulkSimCandidateResult,
	results: ConcurrentBulkSimCandidateResult[],
	iterations: number,
): Promise<{ baseline: ConcurrentBulkSimCandidateResult; results: ConcurrentBulkSimCandidateResult[]; iterations: number }> => {
	const maxAdaptiveIterations = Math.ceil(iterations * BULK_SIM_ADAPTIVE_MAX_ITERATION_MULTIPLIER);
	for (let adaptivePass = 1; adaptivePass <= BULK_SIM_MAX_ADAPTIVE_PASSES; adaptivePass++) {
		if (signals.abort.isTriggered() || hasBulkSimStageError(baseline, results)) return { baseline, results, iterations };

		const observedErrorPct = bulkSimObservedStageErrorPct(baseline, results, iterations, candidates.length);
		if (observedErrorPct <= config.targetErrorPct) return { baseline, results, iterations };

		let targetIterations = getBulkSimStageTargetIterations(config.targetErrorPct, baseline, results, candidates.length);
		targetIterations = Math.min(maxAdaptiveIterations, Math.max(iterations + 1, targetIterations));
		if (targetIterations <= iterations) return { baseline, results, iterations };

		if (isDevMode()) {
			console.log(`[Bulk Sim] - Stage: ${bulkSimStageLogName(config.stage)} - Adaptive pass ${adaptivePass}
Results:
  Current iterations: ${iterations}
  Additional iterations: ${targetIterations - iterations}
  Target iterations: ${targetIterations}
  Target error: ${config.targetErrorPct.toFixed(2)}%
  Observed error: ${observedErrorPct.toFixed(2)}%`);
		}

		const additionalIterations = targetIterations - iterations;
		const rerunResult = await rerunConcurrentBulkSimStageAdditionalIterations(
			request,
			candidates,
			config,
			workerPool,
			onProgress,
			signals,
			baseline,
			results,
			iterations,
			additionalIterations,
		);
		baseline = rerunResult.baseline;
		results = rerunResult.results;
		iterations = targetIterations;
	}
	return { baseline, results, iterations };
};

const rerunConcurrentBulkSimStageAdditionalIterations = async (
	request: BulkSimRequest,
	candidates: ConcurrentBulkSimCandidate[],
	config: ConcurrentBulkSimStageConfig,
	workerPool: WorkerPool,
	onProgress: WorkerProgressCallback,
	signals: SimSignals,
	baseline: ConcurrentBulkSimCandidateResult,
	results: ConcurrentBulkSimCandidateResult[],
	currentIterations: number,
	additionalIterations: number,
): Promise<{ baseline: ConcurrentBulkSimCandidateResult; results: ConcurrentBulkSimCandidateResult[] }> => {
	const totalSims = candidates.length + 1;
	const emitter = makeBulkSimStageProgressEmitter(onProgress, config.stage, totalSims, totalSims * additionalIterations);
	emitter.report(0, 0, 0);

	const baselineExtra = await runBulkSimBaselineSegment(request, baseline, additionalIterations, currentIterations, workerPool, emitter, 0, 0, signals);
	if (baselineExtra.error) return { baseline: baselineExtra, results };
	baseline = baselineExtra;
	emitter.report(1, additionalIterations, baseline.dpsMetrics?.avg ?? 0);

	// The pass so far is the carry-over for this batch, so the extra iterations are merged
	// into each candidate as they complete - exactly like a stage carrying its
	// predecessor's iterations.
	const carried = new Map(results.filter(result => !!result).map(result => [result.candidate.index, result]));
	const merged = await runBulkSimCandidateBatchOnWorkers(
		request,
		candidates,
		additionalIterations,
		workerPool,
		signals,
		{
			completedSimsBase: 1,
			completedIterationsBase: additionalIterations,
			emitter,
			seedOffset: currentIterations,
		},
		carried,
	);

	return { baseline, results: merged };
};

// A stage that could not produce results: the metrics carry only the stage id, since
// nothing was measured.
const bulkSimStageError = (
	config: ConcurrentBulkSimStageConfig,
	baseline: ConcurrentBulkSimCandidateResult,
	iterations: number,
): ConcurrentBulkSimStageResult => ({
	baseline,
	results: [],
	iterations,
	metrics: BulkSimStageMetrics.create({ stage: config.stage }),
});

const buildBulkSimStageMetrics = (
	config: ConcurrentBulkSimStageConfig,
	candidates: ConcurrentBulkSimCandidate[],
	results: ConcurrentBulkSimCandidateResult[],
	baseline: ConcurrentBulkSimCandidateResult,
	iterations: number,
	concurrency: number,
	startedAt: number,
): BulkSimStageMetrics =>
	BulkSimStageMetrics.create({
		stage: config.stage,
		inputGearSets: candidates.length,
		survivors: results.length,
		iterations,
		concurrency,
		durationSeconds: (new Date().getTime() - startedAt) / 1000,
		targetErrorPct: config.targetErrorPct,
		observedErrorPct: bulkSimObservedStageErrorPct(baseline, results, iterations, candidates.length),
		baselineAvgDps: baseline.dpsMetrics?.avg ?? 0,
		bestCandidateAvgDps: topBulkSimResults(results, 1)[0]?.dpsMetrics?.avg ?? 0,
	});

// Iterations already simmed by the previous stage are carried over instead of being
// re-run: the stage only sims the delta with a seed offset and merges. That is exactly
// equivalent to running the full count from scratch, because iteration N always uses
// seed randomSeed+N.
export const runConcurrentBulkSimStage = async (
	request: BulkSimRequest,
	candidates: ConcurrentBulkSimCandidate[],
	config: ConcurrentBulkSimStageConfig,
	workerPool: WorkerPool,
	onProgress: WorkerProgressCallback,
	signals: SimSignals,
	carryOver?: ConcurrentBulkSimStageCarryOver,
): Promise<ConcurrentBulkSimStageResult> => {
	const startedAt = new Date().getTime();
	const minIterations = getBulkSimStageMinIterations(request, config);
	const carry = bulkSimCarryOverCovers(carryOver, candidates) ? carryOver : undefined;
	const carriedIterations = carry?.iterations ?? 0;
	const concurrency = Math.min(workerPool.getNumWorkers(), candidates.length);
	if (isDevMode()) {
		console.log(`[Bulk Sim] ${formatBulkSimStageStart(config, candidates.length, minIterations, carriedIterations)}`);
	}
	const maxBaselineSims = 2;
	const maxTotalSims = candidates.length + maxBaselineSims;
	const probeDelta = Math.max(0, minIterations - carriedIterations);
	const probeEmitter = makeBulkSimStageProgressEmitter(onProgress, config.stage, maxTotalSims, maxTotalSims * probeDelta);
	probeEmitter.report(0, 0, 0);

	const baselineCandidate = { index: -1, gear: getBulkSimBaselineGear(request) };
	let baselineProbe = carry?.baseline;
	let probeIterations = carriedIterations;
	if (probeDelta > 0) {
		baselineProbe = await runBulkSimBaselineSegment(request, baselineProbe, probeDelta, carriedIterations, workerPool, probeEmitter, 0, 0, signals);
		if (baselineProbe.error) {
			return bulkSimStageError(config, baselineProbe, minIterations);
		}
		probeIterations = minIterations;
	}
	// probeDelta is only 0 when the carried iterations already cover the stage minimum,
	// which implies a carried baseline. Guard anyway rather than risk a crash mid-run.
	if (!baselineProbe) {
		const error = ErrorOutcome.create({ message: '[Bulk sim] Stage has no baseline probe' });
		return bulkSimStageError(config, { candidate: baselineCandidate, error }, minIterations);
	}
	probeEmitter.report(1, probeDelta, baselineProbe.dpsMetrics?.avg ?? 0);

	const iterations = Math.max(getBulkSimStageIterations(request, config, baselineProbe.dpsMetrics, candidates.length), probeIterations);
	const baselineDelta = iterations - probeIterations;
	const candidateDelta = iterations - carriedIterations;
	const probeSims = probeDelta > 0 ? 1 : 0;
	const baselineSims = probeSims + (baselineDelta > 0 ? 1 : 0);
	const candidateSims = candidateDelta > 0 ? candidates.length : 0;
	const totalSims = candidateSims + baselineSims;
	let completedBaselineIterations = probeDelta;
	let baseline = baselineProbe;
	const emitter = makeBulkSimStageProgressEmitter(onProgress, config.stage, totalSims, candidateSims * candidateDelta + probeDelta + baselineDelta);
	emitter.report(probeSims, completedBaselineIterations, baselineProbe.dpsMetrics?.avg ?? 0);

	if (baselineDelta > 0) {
		baseline = await runBulkSimBaselineSegment(request, baselineProbe, baselineDelta, probeIterations, workerPool, emitter, probeSims, probeDelta, signals);
		if (baseline.error) {
			return bulkSimStageError(config, baseline, iterations);
		}
		completedBaselineIterations = probeDelta + baselineDelta;
		emitter.report(baselineSims, completedBaselineIterations, baseline.dpsMetrics?.avg ?? 0);
	}

	// Everything this stage needs is already covered by the carried iterations.
	if (candidateDelta <= 0) {
		const carriedResults = bulkSimCarriedResults(carry, candidates);
		return {
			baseline,
			results: carriedResults,
			iterations,
			metrics: buildBulkSimStageMetrics(config, candidates, carriedResults, baseline, iterations, concurrency, startedAt),
		};
	}

	const results = await runBulkSimCandidateBatchOnWorkers(
		request,
		candidates,
		candidateDelta,
		workerPool,
		signals,
		{
			completedSimsBase: baselineSims,
			completedIterationsBase: completedBaselineIterations,
			emitter,
			seedOffset: carriedIterations,
		},
		carry?.byCandidate,
	);
	const adaptedStage = await adaptConcurrentBulkSimStageIterations(
		request,
		candidates,
		config,
		workerPool,
		onProgress,
		signals,
		baseline,
		results,
		iterations,
	);
	baseline = adaptedStage.baseline;
	// Only a rerun produces a fresh array; every other path returns the one passed in, so
	// the common case would copy every element onto itself. Replace in place (the callers
	// below hold this reference) by assignment rather than a spread: the low stage runs on
	// the full candidate list, and one argument per candidate throws RangeError past the
	// engine's argument limit.
	if (adaptedStage.results !== results) {
		for (let i = 0; i < adaptedStage.results.length; i++) {
			results[i] = adaptedStage.results[i];
		}
		results.length = adaptedStage.results.length;
	}
	if (baseline.error) {
		return bulkSimStageError(config, baseline, adaptedStage.iterations);
	}

	return {
		baseline,
		results,
		iterations: adaptedStage.iterations,
		metrics: buildBulkSimStageMetrics(config, candidates, results, baseline, adaptedStage.iterations, concurrency, startedAt),
	};
};
